src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -688,52 +768,97 @@ private static int DoIORequest(ConcurrentQueue<SerialStreamIORequest> q, Request
-                                                           pollWriteEvents: _writeQueue.Count > 0);
+                bool hasPendingReads = _readQueue.Count > 0;
+                bool hasPendingWrites = _writeQueue.Count > 0;
+
+                bool hasPendingIO = hasPendingReads || hasPendingWrites;
+                bool isIdle = _dataReceived == null && !hasPendingIO;
 
-                if (events.HasFlag(Interop.Sys.PollEvents.POLLNVAL) ||
-                    events.HasFlag(Interop.Sys.PollEvents.POLLERR))
+                if (!hasPendingIO)
                 {
-                    // bad descriptor or some other error we can't handle
-                    FinishPendingIORequests();
-                    break;
-                }
+                    if (isIdle)
+                    {
+                        if (!lastIsIdle)
+                        {
+                            // we've just started idling
+                            ticksWhenIdleStarted = Environment.TickCount;
+                        }
+                        else if (Environment.TickCount - ticksWhenIdleStarted > IOLoopIdleTimeout)
+                        {
+                            // we are already idling for a while
+                            // let's stop the loop until there is some work to do
+
+                            lock (_ioLoopLock)
+                            {
+                                // double check we are done under lock
+                                if (_dataReceived == null && _readQueue.Count == 0 && _writeQueue.Count == 0)
+                                {
+                                    _ioLoop = null;
+                                    break;
+                                }
+                                else
+                                {
+                                    // to make sure timer restarts
+                                    lastIsIdle = false;
+                                    continue;
+                                }
+                            }
+                        }
+                    }
 
-                if (events.HasFlag(Interop.Sys.PollEvents.POLLIN))
+                    Thread.Sleep(1);
+                }
+                else
                 {
-                    int bytesRead = DoIORequest(_readQueue, _processReadDelegate);
-                    totalBytesRead += bytesRead;
+                    Interop.Sys.PollEvents events = PollEvents(1,
+                                                               pollReadEvents: hasPendingReads,
+                                                               pollWriteEvents: hasPendingWrites,
+                                                               out Interop.ErrorInfo? error);
+
+                    if (error.HasValue)
+                    {
+                        FinishPendingIORequests(() => Interop.GetIOException(error.Value));
```

Comment:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgUBmLAJgwGEMBvIjRrMlAFgwFkAKASjoaaCAlgDMMXKgHsAdlEkAbOADoASnAgATADJDpcXkq1xpAcxgALDAEIAvBnI8BgxvULP3GXTAwAPDHa8lAAUIBCh9KVkFZTVNHT1eHgI3D0EUAFYAHi8APgwxO15/PL8AKntk1LTsAE4uEUTKjwBfJwxWwmagA==

says that it generates the code but the allocation happens only in the if statement - . Is there any easy way to remove this allocation and code gen? The only option I can think of ATM is passing `Type` and `object[]` and then calling `Activator.CreateInstance` but that would require allocating the array and using reflection feels like even worse solution I assume sharing Exception instance is not an option, right?

Reply:
> says that it generates the code but the allocation happens only in the if statement

That's not an equivalent case.  Move the declaration of `x` to outside of the `if` and check again.

Reply:
@stephentoub - ahh right.. I guess I will just duplicate the code since I'm not seeing any easy solution

Reply:
> ahh right.. I guess I will just duplicate the code since I'm not seeing any easy solution

Why not just change FinishPendingIORequests to accept the Error as another argument?  You could then avoid any closure/delegate allocation here by having that value passed into the delegate (which you'd define to accept an argument as well).

Or if you're ok with the allocations here on the error path, just copy the value into a local in the closest scope and then capture that local:
```C#
Interop.ErrorInfo localError = error.GetValueOrDefault();
FinishPendingIORequests(() => Interop.GetIOException(localError));
```

Reply:
thought about this solution after already sending new iteration - I think I like the solution with less code gen in your latest suggestion more so will do that instead - got no strong opinion here.